src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,166 @@
+        );
+      case 'site':
+        if (tooMany) {
+          return this.i18n.sprintf(this.i18n.ngettext(
+            'Access your data on %(param)s other site',
+            'Access your data on %(param)s other sites',
+            param), { param: this.i18n.formatNumber(param) }
+          );
+        }
+        return this.i18n.sprintf(
+          this.i18n.gettext('Access your data for %(param)s'),
+          { param }
+        );
+      default:
+        return '';
+    }
+  }
+
+  i18n: I18nType;
+
+  // Generates Permission components for a list of host permissions. If we have 4 or
+  // fewer, display them all, otherwise display the first 3 followed by an item
+  // that says "...plus N others".
+  generateHostPermissions({
+    permissions, messageType,
+  }: GenerateHostPermissionsParams): Array<React.Element<typeof Permission>> {
+    const hostPermissions = [];
+    for (const item of permissions.slice(0, 4)) {
+      // Add individual Permission components for the first 4 host permissions.
+      hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.getPermissionString({ messageType, param: item })}
+          key={item}
+        />
+      );
+    }
+    if (permissions.length > 4) {
+      // Replace the final individual permission with a "too many" permission.
+      hostPermissions[3] = (<Permission
+        type="hostPermission"
+        description={this.getPermissionString(
+          { messageType, param: permissions.length - 3, tooMany: true }
+        )}
+        key={messageType}
+      />);
+    }
+    return hostPermissions;
+  }
+
+  render() {
+    const { permissions } = this.props;
+    const hostPermissions = [];
+
+    // Group permissions into "site" and "domain" permissions. If any
+    // "all urls" permissions are found, break as we'll only need one
+    // host permissions message.
+    let allUrls = false;
+    const wildcards = [];
+    const sites = [];
+    for (const permission of permissions) {
+      if (permission === '<all_urls>') {
+        allUrls = true;
+        break;
+      }
+      if (permission.startsWith('moz-extension:')) {
+        continue;
+      }
+      const match = /^[a-z*]+:\/\/([^/]+)\//.exec(permission);
+      if (!match) {
+        log.debug(`Host permission string "${permission}" appears to be invalid.`);
+        continue;
+      }
+      if (match[1] === '*') {
+        allUrls = true;
+      } else if (match[1].startsWith('*.')) {
+        wildcards.push(match[1].slice(2));
+      } else {
+        sites.push(match[1]);
+      }
+    }
+
+    // Format the host permissions. If we have a wildcard for all urls,
+    // a single string will suffice.  Otherwise, show domain wildcards
+    // first, then individual host permissions.
+    if (allUrls) {
+      hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.getPermissionString({ messageType: 'allUrls' })}
```

Comment:
I guess it's fine because you have flow but I always prefer things like this as constants. It makes it easier to change their values later... but it's up to you.

Reply:
I agree that it's better to have constants. Flow seems to both protect us without them and also require extra code with them, but the latter is better, so I've made the change. :)